sdks/go/pkg/beam/core/runtime/harness/statemgr_test.go

```diff
@@ -258,6 +261,167 @@ func TestStateChannel(t *testing.T) {
+			buflens:  []int{-1, readLen, readLen},
+			numReads: 4,
+		}, {
+			name:     "emptyDataMid", // Shouldn't happen, but not unreasonable to handle.
+			buflens:  []int{readLen, readLen, -1, readLen},
+			numReads: 5,
+		}, {
+			name:     "emptyDataLast", // Shouldn't happen, but not unreasonable to handle.
+			buflens:  []int{readLen, readLen, -1},
+			numReads: 3,
+		}, {
+			name:     "emptyDataLast-short",
+			buflens:  []int{3*readLen - 2, -1},
+			numReads: 4,
+		}, {
+			name:     "closed",
+			buflens:  []int{-1, -1},
+			numReads: 1,
+			closed:   true,
+		}, {
+			name:     "noGet",
+			buflens:  []int{-1},
+			numReads: 1,
+			noGet:    true,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctx, cancelFn := context.WithCancel(context.Background())
+			ch := &StateChannel{
+				id:        "test",
+				requests:  make(chan *fnpb.StateRequest),
+				responses: make(map[string]chan<- *fnpb.StateResponse),
+				cancelFn:  cancelFn,
+				DoneCh:    ctx.Done(),
+			}
+
+			// Handle the channel behavior asynchronously.
+			go func() {
+				for i, buflen := range test.buflens {
+					token := []byte(strconv.Itoa(i))
+					var buf []byte
+					if buflen >= 0 {
+						buf = bytes.Repeat([]byte{42}, buflen)
+					}
+					// On the last request response pair, send no token.
+					if i+1 == len(test.buflens) {
+						token = nil
+					}
+
+					req := <-ch.requests
+
+					if test.noGet {
+						ch.responses[req.Id] <- &fnpb.StateResponse{
+							Id: req.Id,
+						}
+						return
+					}
+
+					ch.responses[req.Id] <- &fnpb.StateResponse{
+						Id: req.Id,
+						Response: &fnpb.StateResponse_Get{
+							Get: &fnpb.StateGetResponse{
+								ContinuationToken: token,
+								Data:              buf,
+							},
+						},
+					}
+				}
+			}()
+
+			r := stateKeyReader{
+				ch: ch,
+			}
+
+			if test.closed {
+				err := r.Close()
+				if err != nil {
+					t.Errorf("unexpected error on Close(), got %v", err)
+				}
+			}
+
+			var totalBytes int
+			for _, l := range test.buflens {
+				if l > 0 {
+					totalBytes += l
+				}
+			}
+			var finalerr error
+			var count, reads int
+
+			// Read all the bytes.
+			for count <= totalBytes {
+				reads++
+				b := make([]byte, readLen) // io.Read is keyed off of length, not capacity.
+				n, err := r.Read(b)
+				if err != nil {
+					finalerr = err
+					break
+				}
+				count += n
+				// Special check to avoid spurious zero elements.
+				if count == totalBytes && n == 0 {
+					t.Error("expected byte count read, last read is 0, but no EOF")
+				}
+			}
+			if got, want := reads, test.numReads; got != want {
+				t.Errorf("read %d times, want %d", got, want)
+			}
+			if got, want := count, totalBytes; got != want {
+				t.Errorf("read %v bytes, want %v", got, want)
+			}
+			if test.closed {
+				if got, want := finalerr, errors.New("side input closed"); !contains(got, want) {
+					t.Errorf("got err %v, want to contain %v", got, want)
```

Comment:
Nit: This error message could benefit from some escaped quotation marks around the %v, to make it explicit where the quoted error begins and ends. Alternatively, surrounding the quoted error with some newlines and tabs might work well too, like so:

```
got err:
    <error message>
want to contain:
    <error message>
```

Reply:
I don't like adding newlines to the test formatted output, since it messes with the indentation of the results, possibly making it harder to read. I have used %q which will quote and escape the Error() strings though.